### PR TITLE
SharedTree: Editable-tree subtreeChanging and PathVisitor

### PR DESCRIPTION
**Pull request by dstanesc**
## Description

Follow-up to PR [15097](https://github.com/microsoft/FluidFramework/pull/15097). SharedTree [data binder](https://github.com/microsoft/FluidFramework/issues/14293) related.

- Eliminates the redundancy between `subtreeChanging` and `visitSubtreeChanging` anchor events as [suggested](https://github.com/microsoft/FluidFramework/pull/15097#discussion_r1171858823) by @noencke 
- Enhances editable tree events with equivalent `subtreeChanging` event returning now `PathVisitor | void`
- Adds utility function that checks if an object implements the `PathVisitor` interface - `isPathVisitor(obj: any)`
- Test updates

## Breaking Changes
Usage of `AnchorEvents.visitSubtreeChanging` is discontinued.

FYI @noencke @CraigMacomber @sharptrip

### CODE REVIEW

File: experimental/dds/tree2/src/core/tree/anchorSet.ts [hunk 1]
```diff
@@ -646,6 +639,7 @@ export class AnchorSet implements ISubscribable<AnchorSetRootEvents> {
 				assert(parent !== undefined, 0x5e9 /* Must be in a node to set its value */);
 				for (const visitors of pathVisitors.values()) {
 					for (const pathVisitor of visitors) {
+						console.log("onSetValue", typeof pathVisitor);
```
Comment (noencke) [inline experimental/dds/tree2/src/core/tree/anchorSet.ts:642]: Remove, leftover from debugging

File: experimental/dds/tree2/src/core/tree/anchorSet.ts [hunk 2]
```diff
@@ -657,13 +651,16 @@ export class AnchorSet implements ISubscribable<AnchorSetRootEvents> {
...
-							"visitSubtreeChanging",
+						const visitors: (PathVisitor | void)[] = p.events.emitAndCollect(
+							"subtreeChanging",
 							p,
 						);
-						if (visitors.length > 0) pathVisitors.set(p, visitors);
+						if (visitors.length > 0)
+							pathVisitors.set(
+								p,
+								visitors.filter((v): v is PathVisitor => isPathVisitor(v)),
```
Comment (noencke) [inline experimental/dds/tree2/src/core/tree/anchorSet.ts:661]: `void` === `undefined`, so it should be type-safe to write this as `(v): v is PathVisitor => v !== undefined`. Then you would not need to have the type guard function `isPathVisitor`.

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableTreeTypes.ts [hunk 3]
```diff
@@ -99,11 +100,13 @@ export interface EditableTreeEvents {
 
 	/**
 	 * Raised when something in the tree is changing, including this node and its descendants.
+	 * The event can return {@link PathVisitor} traversing the subtree or `void`
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/editable-tree/editableTreeTypes.ts:103]:
```suggestion
	 * The event can optionally return a {@link PathVisitor} to traverse the subtree
```

File: experimental/dds/tree2/src/core/tree/visithPath.ts
Comment (noencke) [inline experimental/dds/tree2/src/core/tree/visithPath.ts:1]: If it's not too much trouble, can this file be renamed to `visitPath`? This typo snuck by in the last PR.

File: experimental/dds/tree2/src/core/tree/anchorSet.ts [hunk 4]
```diff
@@ -114,18 +114,11 @@ export interface AnchorEvents {
 
 	/**
 	 * Something in this tree is changing.
+	 * The event can return {@link PathVisitor} traversing the subtree or `void`
```
Comment (noencke) [inline experimental/dds/tree2/src/core/tree/anchorSet.ts:117]:
```suggestion
	 * The event can optionally return {@link PathVisitor} to traverse the subtree
```

File: experimental/dds/tree2/src/core/tree/anchorSet.ts (hunk 1 above)
Comment (dstanesc) [reply experimental/dds/tree2/src/core/tree/anchorSet.ts]: Done

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableTreeTypes.ts (hunk 3 above)
Comment (dstanesc) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableTreeTypes.ts]: Done

File: experimental/dds/tree2/src/core/tree/visithPath.ts
Comment (dstanesc) [reply experimental/dds/tree2/src/core/tree/visithPath.ts:1]: Not at all. Done.

File: experimental/dds/tree2/src/core/tree/anchorSet.ts (hunk 2 above)
Comment (dstanesc) [reply experimental/dds/tree2/src/core/tree/anchorSet.ts]: Done. This is where I started :) but somehow VSCode convinced me that it is insufficient. Now everything looks clean! Started to believe that linting lags or inconsistencies are sometimes present. Learning. Thank you!

**Review by noencke — APPROVED** [review]

**Pull request merged**